tokio-trace-futures/examples/proxy_server.rs

```diff
@@ -0,0 +1,153 @@
+    println!("Listening on: {}", listen_addr);
+    println!("Proxying to: {}", server_addr);
+
+    let done = socket
+        .incoming()
+        .map_err(|e| debug!(msg = "error accepting socket", error = field::display(e)))
+        .for_each(move |client| {
+            let server = TcpStream::connect(&server_addr);
+            let mut client_addr = String::new();
+            match client.peer_addr() {
+                Ok(x) => {
+                    let x = x.to_string();
+                    client_addr = x;
+                    info!(
+                        message = "client connected",
+                        client_addr = field::display(&client_addr)
+                    )
+                }
+                Err(e) => debug!(
+                    message = "could not get client info",
+                    error = field::display(e)
+                ),
+            }
+
+            let amounts = server.and_then(move |server| {
+                // Create separate read/write handles for the TCP clients that we're
+                // proxying data between. Note that typically you'd use
+                // `AsyncRead::split` for this operation, but we want our writer
+                // handles to have a custom implementation of `shutdown` which
+                // actually calls `TcpStream::shutdown` to ensure that EOF is
+                // transmitted properly across the proxied connection.
+                //
+                // As a result, we wrap up our client/server manually in arcs and
+                // use the impls below on our custom `MyTcpStream` type.
+                let client_reader = MyTcpStream(Arc::new(Mutex::new(client)));
+                let client_writer = client_reader.clone();
+                let server_reader = MyTcpStream(Arc::new(Mutex::new(server)));
+                let server_writer = server_reader.clone();
+
+                // Copy the data (in parallel) between the client and the server.
+                // After the copy is done we indicate to the remote side that we've
+                // finished by shutting down the connection.
+                let client_to_server = copy(client_reader, server_writer)
+                    .and_then(|(n, _, server_writer)| {
+                        info!(size = n);
+                        shutdown(server_writer).map(move |_| n)
+                    })
+                    .instrument(span!("client_to_server"));
+
+                let server_to_client = copy(server_reader, client_writer)
+                    .and_then(|(n, _, client_writer)| {
+                        info!(size = n);
+                        shutdown(client_writer).map(move |_| n)
+                    })
+                    .instrument(span!("server_to_client"));
+
+                client_to_server.join(server_to_client)
+            });
+
+            let msg = amounts
+                .map(move |(from_client, from_server)| {
+                    info!(
```

Comment:
Take it or leave it, but it might also be worth including a message (maybe "transfer completed"?) in this event, as well as the amounts of data transferred...

Reply:
Done